### PR TITLE
Feature | tasks

### PR DESCRIPTION
**Pull request by WoLkErSs**

### CODE REVIEW

File: app/controllers/api/v1/tasks_controller.rb [hunk 1]
```diff
@@ -0,0 +1,37 @@
+class Api::V1::TasksController < ApplicationController
+  before_action :authorize_request
+  load_and_authorize_resource :project, only: :create
+  load_and_authorize_resource :task, through: :project, only: :create
+  load_and_authorize_resource except: :create
+
+  def index
+    render_correct_json(@tasks.order('position DESC'))
```
Comment (za4emnik) [inline app/controllers/api/v1/tasks_controller.rb]: do we have not correct?

File: app/controllers/api/v1/tasks_controller.rb [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    UpdateTaskService.new.call(@task, task_params) ? render_correct_json(@task) : bad_request
+  end
+
+  def destroy
+    @task.destroy
+  end
+
+  private
+
+  def render_correct_json(entity)
```
Comment (za4emnik) [inline app/controllers/api/v1/tasks_controller.rb]: can we call dynamically serializers?

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -9,7 +9,7 @@ def not_found
   end
 
   def bad_request
-    render json: { error: 'Bad Request' }, status: :bad_request
+    render json: { error: 'bad_request' }, status: :bad_request
```
Comment (za4emnik) [inline app/controllers/application_controller.rb]: maybe need to show an error message?

File: app/models/task.rb [hunk 4]
```diff
@@ -2,6 +2,17 @@
...
-  validates :name, presence: true
+  validates :deadline, presence: true
+  validates :name,
+            length: { minimum: 3, maximum: 50 },
+            presence: true
+
+  validate :deadline_in_future
+
+  def deadline_in_future
+    errors.add(:deadline, ' should be future time') if Time.zone.now.strftime('%Y-%m-%d').to_date > deadline.to_date
```
Comment (za4emnik) [inline app/models/task.rb]: i18n?

File: spec/models/coment_spec.rb [hunk 5]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
```
Comment (za4emnik) [inline spec/models/coment_spec.rb]: ?

File: spec/requests/v1/projects_spec.rb [hunk 6]
```diff
@@ -94,9 +94,9 @@
     end
 
     it 'delete the project', :dox do
-      patch "/api/v1/projects/#{project.id}", params: valid_params, headers: valid_headers, as: :json
-      expect(response).to match_json_schema('projects/delete')
-      expect(response).to have_http_status(200)
+      delete "/api/v1/projects/#{project.id}", headers: valid_headers, as: :json
```
Comment (za4emnik) [inline spec/requests/v1/projects_spec.rb]: subject

File: spec/requests/v1/tasks_spec.rb [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+  let(:token) { JsonWebTokenService.encode(user_id: task.project.user_id) }
+  let(:valid_headers) { { 'Content-Type': 'application/json', 'Authorization': token} }
+  let(:invalid_params) { { task: { name: '', deadline: '2000-08-29' } } }
+  let(:valid_params) { { task: { name: FFaker::CheesyLingo.title, deadline: (Time.zone.now + 2.days).strftime('%Y-%m-%d') } } }
+
+  describe 'GET /api/v1/projects/{project_id}/tasks' do
+    include Docs::V1::Tasks::Index
+
+    it 'with correct token', :dox do
+      get "/api/v1/projects/#{task.project_id}/tasks", headers: valid_headers
```
Comment (za4emnik) [inline spec/requests/v1/tasks_spec.rb]: subject

File: spec/requests/v1/tasks_spec.rb [hunk 8]
```diff
@@ -0,0 +1,132 @@
+RSpec.describe 'V1::Tasks API', type: :request do
+  include Docs::V1::Tasks::Api
+
+  let!(:task) { create(:task) }
+  let(:token) { JsonWebTokenService.encode(user_id: task.project.user_id) }
+  let(:valid_headers) { { 'Content-Type': 'application/json', 'Authorization': token} }
+  let(:invalid_params) { { task: { name: '', deadline: '2000-08-29' } } }
```
Comment (za4emnik) [inline spec/requests/v1/tasks_spec.rb]: can we use `attributes_for`?

File: spec/requests/v1/tasks_spec.rb [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+      expect(response).to match_json_schema('tasks/create')
+      expect(response).to have_http_status(200)
+    end
+  end
+
+  describe 'PATCH /api/v1/tasks/{task_id}' do
+    include Docs::V1::Tasks::Update
+
+    let(:new_status_params) { { task: { done: 'true' } } }
+    let(:valid_position_params) { { task: { position: 'up' } } }
```
Comment (za4emnik) [inline spec/requests/v1/tasks_spec.rb]: use `let` and redefine params in `context`

File: app/controllers/application_controller.rb [hunk 10]
```diff
@@ -4,22 +4,34 @@ class ApplicationController < ActionController::API
...
   def current_user
     @current_user ||= User.find(JsonWebTokenService.decode(token)[:user_id])
   end
 
+  def render_success_response(entity)
+    render json: "#{entity.class.name}Serializer".constantize.new(entity).serialized_json, status: :ok
+  end
+
   def authorize_request
     @decoded = JsonWebTokenService.decode(token)
```
Comment (za4emnik) [inline app/controllers/application_controller.rb]: decoded?

File: app/controllers/application_controller.rb [hunk 11]
```diff
@@ -4,22 +4,34 @@ class ApplicationController < ActionController::API
...
     @current_user = User.find(@decoded[:user_id])
```
Comment (za4emnik) [inline app/controllers/application_controller.rb]: we set current user here?

File: app/controllers/application_controller.rb [hunk 12]
```diff
@@ -4,22 +4,34 @@ class ApplicationController < ActionController::API
...
 
   def current_user
     @current_user ||= User.find(JsonWebTokenService.decode(token)[:user_id])
   end
 
+  def render_success_response(entity)
+    render json: "#{entity.class.name}Serializer".constantize.new(entity).serialized_json, status: :ok
+  end
+
   def authorize_request
```
Comment (za4emnik) [inline app/controllers/application_controller.rb]: separate this method

File: app/models/ability.rb [hunk 13]
```diff
@@ -7,5 +7,6 @@ def initialize(user)
     user ||= User.new
     can :manage, Project, user_id: user.id
     can :manage, Task, project: { user_id: user.id }
+    can %i[create read destroy], Comment, task: { project: { user_id: user.id } }
```
Comment (za4emnik) [inline app/models/ability.rb]: `:crud`?

File: app/models/task.rb [hunk 14]
```diff
@@ -2,6 +2,19 @@
...
+  validates :deadline, presence: true
+  validates :name,
+            length: { minimum: 3, maximum: 50 },
+            presence: true
+
+  validate :deadline_in_future
+
+  def deadline_in_future
+    errors.add(:deadline, I18n.t('tasks.wrong_time_param')) if Time.zone.now.strftime('%Y-%m-%d').to_date > deadline.to_date
+  rescue NoMethodError
```
Comment (za4emnik) [inline app/models/task.rb]: for what?

File: app/services/update_task_service.rb [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+
+  private
+
+  def update_task(task, params)
+    task.update(name: params[:name], deadline: params[:deadline])
+  end
+
+  def update_position(task, params)
+    case params[:position]
+    when 'up' then task.move_lower
```
Comment (za4emnik) [inline app/services/update_task_service.rb]: constant

File: spec/models/coment_spec.rb [hunk 16]
```diff
@@ -0,0 +1,5 @@
...
+
+RSpec.describe Comment, type: :model do
```
Comment (za4emnik) [inline spec/models/coment_spec.rb]: it really resolved?

File: spec/requests/v1/tasks_spec.rb [hunk 17]
```diff
@@ -0,0 +1,221 @@
+RSpec.describe 'V1::Tasks API', type: :request do
+  include Docs::V1::Tasks::Api
+
+  let!(:task) { create(:task) }
+  let(:token) { JsonWebTokenService.encode(user_id: task.project.user_id) }
+  let(:valid_headers) { { 'Content-Type': 'application/json', 'Authorization': token } }
+  let(:no_token_headers) { { 'Content-Type': 'application/json' } }
+  let(:invalid_params) { { task: FactoryBot.attributes_for(:task, name: nil, deadline: '2000-08-29') } }
+  let(:valid_params) { { task: { name: FFaker::CheesyLingo.title, deadline: (Time.zone.now + 2.days).strftime('%Y-%m-%d') } } }
```
Comment (za4emnik) [inline spec/requests/v1/tasks_spec.rb]: use factory

File: spec/requests/v1/tasks_spec.rb [hunk 18]
```diff
@@ -0,0 +1,221 @@
...
+  describe 'GET /api/v1/projects/{project_id}/tasks' do
+    include Docs::V1::Tasks::Index
+
+    subject{ get "/api/v1/projects/#{task.project_id}/tasks", headers: header }
+
+    context 'when show tasks with correct a token' do
+      let(:header) { valid_headers }
+
+      it 'with correct token', :dox do
+        subject
```
Comment (za4emnik) [inline spec/requests/v1/tasks_spec.rb]: can we use `before?`

File: spec/requests/v1/tasks_spec.rb [hunk 19]
```diff
@@ -0,0 +1,221 @@
...
+      let(:header) { no_token_headers }
+
+      it 'unauthorized', :dox do
+        subject
+        expect(response).to have_http_status(:unauthorized)
+        expect(response).to match_json_schema('errors/unauthorized')
+      end
+    end
+
+    context 'whne show a task by token' do
```
Comment (za4emnik) [inline spec/requests/v1/tasks_spec.rb]: when

File: spec/models/coment_spec.rb (hunk 16 above)
Comment (WoLkErSs) [reply spec/models/coment_spec.rb]: 'Tasks' branch has specs only for tasks_controller

File: app/models/ability.rb (hunk 13 above)
Comment (WoLkErSs) [reply app/models/ability.rb]: no, because nobody can update a comment

File: app/services/create_task_service.rb [hunk 20]
```diff
@@ -0,0 +1,9 @@
+class CreateTaskService
+  def call(task, params, project)
+    position = project.tasks.count + 1
```
Comment (za4emnik) [inline app/services/create_task_service.rb]: ++

File: config/environments/development.rb [hunk 21]
```diff
@@ -14,6 +14,8 @@
   # Show full error reports.
   config.consider_all_requests_local = true
 
+  # config.cache_store = :redis_cache_store, { url: 'redis://localhost:3000/1' }
```
Comment (za4emnik) [inline config/environments/development.rb]: ?

File: spec/requests/v1/tasks_spec.rb [hunk 22]
```diff
@@ -0,0 +1,208 @@
...
+        expect(response).to match_json_schema('tasks/update')
+        expect(response).to have_http_status(200)
+      end
+    end
+
+    context 'when change status' do
+      let(:params) { { task: { done: 'true' } } }
+      before { subject }
+
+      it 'with new status', :dox do
```
Comment (za4emnik) [inline spec/requests/v1/tasks_spec.rb]: with?

File: spec/requests/v1/tasks_spec.rb [hunk 23]
```diff
@@ -0,0 +1,208 @@
...
+        expect(response).to match_json_schema('tasks/update')
+        expect(response).to have_http_status(200)
+      end
+    end
+
+    context 'when can not change position' do
+      let(:params) { { task: { position: 'up' } } }
```
Comment (za4emnik) [inline spec/requests/v1/tasks_spec.rb]: use const

File: app/services/create_task_service.rb (hunk 20 above)
Comment (WoLkErSs) [reply app/services/create_task_service.rb]: ```Ruby has no pre/post increment/decrement operator```

**Pull request merged**